### DISCUSSION TITLE
Cygwin demos/guide fails: create_socket_bio fails - library:BIO_connect:Connection refused:crypto/bio/bio_sock2.c:114:calling connect()

### DISCUSSION THREAD

**Discussion by gidigal** [openssl/openssl#26017] [Q&A] [answered]
My Cygwin version:
$ uname -a
CYGWIN_NT-10.0-19045 DESKTOP-PH67H23 3.5.4-1.x86_64 2024-08-25 16:52 UTC x86_64 Cygwin

openssl version: 3.4.0

I am running server:
$ LD_LIBRARY_PATH=../.. ../../apps/openssl s_server -www -accept localhost:4443 -cert servercert.pem -key serverkey.pem
I am running client (after running make):
$ SSL_CERT_FILE=rootcert.pem LD_LIBRARY_PATH=../.. ./tls-client-block localhost 4443
Failed to crete the BIO
100000000A000000:error:8000006F:system library:BIO_connect:Connection refused:crypto/bio/bio_sock2.c:114:calling connect()
100000000A000000:error:10000067:BIO routines:BIO_connect:connect error:crypto/bio/bio_sock2.c:116:

I tried looking for the port status:
C:\WINDOWS\system32>netstat -na | find "4443"
  TCP    [::1]:4443             [::]:0                 LISTENING

**Comment by Sashan** [comment]
Any chance the connection from `tls-client-block` application is blocked by firewall? the error seems to be coming from call to connect provided by windows runtime library.

**Reply by gidigal** [reply]
At first, I could see in the firewall GUI that my application is blocked and I got also "Permission denied" error. Now, after re-installing the firewall I don't see anymore reports for blocking the client and server. However, I still get "Connection refused" error as displayed in my original report.

**Comment by nhorman** [comment]
your client is trying to connect via ipv4, while your server is listening on ipv6 addreses.  Try running the client like this:
```
SSL_CERT_FILE=rootcert.pem LD_LIBRARY_PATH=../.. ./tls-client-block localhost -6 4443
```

**Comment by gidigal** [comment]
SSL_CERT_FILE=rootcert.pem LD_LIBRARY_PATH=../.. ./tls-client-block localhost -6 4443
Failed to crete the BIO
100000000A000000:error:8000006F:system library:BIO_connect:Connection refused:crypto/bio/bio_sock2.c:114:calling connect()
100000000A000000:error:10000067:BIO routines:BIO_connect:connect error:crypto/bio/bio_sock2.c:116:

**Comment by mattcaswell** [comment]
I think the command @nhorman meant was actually this (note the ordering of the params):

````
SSL_CERT_FILE=rootcert.pem LD_LIBRARY_PATH=../.. ./tls-client-block -6 localhost 4443
````

**Reply by gidigal** [reply]
Thanks @mattcaswell  and @nhorman  for your help. I managed to fix my code regarding this call, now it is stuck on the call to SSL_connect. My goal is to use SSL in cpp application. I am following the example in tls-client-block. I created a class named TLSBlockingClient (attached) for this purpose. I will be grateful if you could take a look and comment what is missing to pass this call correctly. The call to SSL_connet is in function  TLSBlockingClient::serverHandshake(). Thanks !
[src.zip](https://github.com/user-attachments/files/18309656/src.zip)

**Reply by nhorman** [reply]
Did you look at the errors your client produces?

```
nhorman@hmsbeagle:~/Downloads/test$ SSL_CERT_FILE=~/git/openssl/test/certs/rootcert.pem LD_LIBRARY_PATH=../.. ./testclient -6 localhost 4443
IPV6 !
Server: localhost port: 4443 family: AF_INET6
loop 0x18460a60
sock value 3
serverHandshake SSL_set_bio
serverHandshake SSL_set_tlsext_host_name
serverHandshake SSL_set1_host hostName: localhost
serverHandshake SSL_connect
connectRes -1
Connect error: 1
Verify error !!!
Failed to connect to the server Verify error: hostname mismatch
Error shutting down
```
The connect fails because the hostname you are connecting to (localhost) doesn't match the hostname specified in the certificate you are using, which:
```
openssl x509 -in servercert.pem -text -noout
```
shows to be server.example
you should create a certificate for the server which contains localhost as the subject CN or server alternate name.  Alternatively you can add server.example as an alias for localhost in your /etc/hosts file for testing purposes.

**Reply by gidigal** [reply, marked as answer]
Thanks @nhorman , I did not see the output you included, the call to SSL_connect was blocked. I tried to look in https://docs.openssl.org/ for some information regarding creation of certificates and could not find one. So, I tried following the instructions in https://www.bastionxp.com/blog/how-to-create-self-signed-ssl-tls-x.509-certificates-using-openssl.
I ran the server with -cert pointing to server-cert.pem and -key pointing to server-key.pem (in both client and server I set "localhost" as my CN). I ran my client with SSL_CERT_FILE pointing to client-cert.pem.
The output:
IPV6 !
Server: localhost port: 4443 family: AF_INET6
loop 0xa000ae540
sock value 4
serverHandshake SSL_set_bio
serverHandshake SSL_set_tlsext_host_name
serverHandshake SSL_set1_host hostName: localhost
serverHandshake SSL_connect
connectRes -1
Connect error: 1
Verify error !!!
Failed to connect to the server Verify error: unable to get local issuer certificate
Error shutting down

Attached are the certificates I created. I found out that my ca-cert.pem did not have expiration date set correctly at previous attempt. I also found out that I was not working with the built openssl library code that I downloaded, but with cygwin's installed openssl, so I updated my $PATH variable to make sure I am working with the built one.
[certificates.zip](https://github.com/user-attachments/files/18343715/certificates.zip)

UPDATE
=====
I managed to fix the issues I had.
1) The root CA certitifcate was with a single day expiration, eventhough I stated "-days 365" while creating it.
I found out that the openssl I used in Cygwin was the one installed when Cygwin was installed and not the openssl built from the source code I downloaded. I added the path to openssl to bash_profile (the file that Cygwin runs when terminal is launched). I also added "-CAstore" switch in my call to launch the server. I think this fixed the expiration date issue and also the "unable to get local issuer certificate".
2) I tried to make sure that my server returns a correct certificate chain, I used this call:
openssl s_client -connect <host:port> -showcerts -CAfile <path to root CA certificate file> -servername <the name of my server host. localhost in my case>
The "-servername localhost" avoids the message "Can't use SSL_get_servername".
Running this call revealed error "verify error:num=19:self-signed certificate in certificate chain". Additional research revealed that I must install my self-signed root ca certificate file in order for it to be trusted. I used this link for this task:
https://docs.devolutions.net/dgw/kb/general-knowledge/install-root-certificates/
3) I tried running my code again and SSL_connect returned -1 with error "Verify error: self-signed certificate in certificate chain".
I found out that I needed to load the root CA file with a call to SSL_CTX_load_verify_locations (after calling SSL_CTX_new).
4) Now I got a response from the server. However, at its end I saw the line "no client certificate available".
This was solved by adding calls to SSL_CTX_use_certificate_file (passing client certificate file), SSL_CTX_use_PrivateKey_file (passing client key file), and SSL_CTX_check_private_key (passing pointer to the CTX object). Also, on server side the switch "-Verify 1" was added. The value is maximum depth for the certificate chain verification. My chain is simple, so depth 1 is enough.
